crml/nft/src/lib.rs

```diff
@@ -974,6 +999,120 @@ decl_module! {
 				None => Err(Error::<T>::NotForFixedPriceSale.into()),
 			}
 		}
+
+		/// Create an offer on a token
+		/// Locks funds until offer is accepted, rejected or cancelled
+		#[weight = T::WeightInfo::make_simple_offer()]
+		#[transactional]
+		fn make_simple_offer (
+			origin,
+			token_id: TokenId,
+			amount: Balance,
+			asset_id: AssetId,
+			marketplace_id: Option<MarketplaceId>,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			ensure!(!amount.is_zero(), Error::<T>::ZeroOffer);
+			ensure!(Self::token_owner((token_id.0, token_id.1), token_id.2) != origin, Error::<T>::IsTokenOwner);
+			let offer_id = Self::next_offer_id();
+			ensure!(offer_id.checked_add(One::one()).is_some(), Error::<T>::NoAvailableIds);
+
+			// ensure the token_id is not currently in an auction
+			if let Some(TokenLockReason::Listed(listing_id)) = Self::token_locks(token_id) {
+				match Self::listings(listing_id) {
+					Some(Listing::<T>::Auction(_)) => return Err(Error::<T>::TokenOnAuction.into()),
+					None | Some(Listing::<T>::FixedPrice(_)) => (),
+				}
+			}
+			// check user has the required funds to make this offer
+			let balance = T::MultiCurrency::free_balance(&origin, asset_id);
+			if let Some(balance_after_bid) = balance.checked_sub(amount) {
+				// TODO: review behaviour with 3.0 upgrade: https://github.com/cennznet/cennznet/issues/414
+				// - `amount` is unused
+				// - if there are multiple locks on user asset this could return true inaccurately
+				// - `T::MultiCurrency::reserve(origin, asset_id, amount)` should be checking this internally...
+				let _ = T::MultiCurrency::ensure_can_withdraw(&origin, asset_id, amount, WithdrawReasons::RESERVE, balance_after_bid)?;
+			}
+
+			// try lock funds
+			T::MultiCurrency::reserve(&origin, asset_id, amount)?;
+			TokenOffers::append(token_id, offer_id);
+			let new_offer = OfferType::<T::AccountId>::Simple(
+				SimpleOffer{
+					token_id,
+					asset_id,
+					amount,
+					buyer: origin.clone(),
+					marketplace_id,
+				}
+			);
+			Offers::<T>::insert(offer_id, new_offer);
+			NextOfferId::mutate(|i| *i += 1);
+
+			Self::deposit_event(RawEvent::OfferMade(offer_id, amount, asset_id, marketplace_id, origin));
+			Ok(())
+		}
+
+		/// Cancels an offer on a token
+		/// Caller must be the offer buyer
+		#[weight = T::WeightInfo::cancel_offer()]
+		fn cancel_offer (
+			origin,
+			offer_id: OfferId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			if let Some(offer_type) = Self::offers(offer_id) {
+				match offer_type {
+					OfferType::Simple(offer) => {
+						ensure!(offer.buyer == origin, Error::<T>::NotBuyer);
+						T::MultiCurrency::unreserve(&origin, offer.asset_id, offer.amount);
+						Offers::<T>::remove(offer_id);
+						TokenOffers::remove(offer.token_id);
```

Comment:
does this remove the element _at_ index `offer.token_id` or does it remove the element _with_ value `offer.token_id`?

Reply:
Good catch! It should be the same as accepting an offer. I'll change it and add some tests in for it